core/src/test/java/org/apache/iceberg/TestLocationProvider.java

```diff
@@ -0,0 +1,164 @@
+    }
+  }
+
+  // publicly visible for testing to be dynamically loaded
+  public static class InvalidDynamicallyLoadedLocationProvider implements LocationProvider {
+    // No public constructor
+
+    @Override
+    public String newDataLocation(String filename) {
+      throw new IllegalStateException("Should have never been instantiated");
+    }
+
+    @Override
+    public String newDataLocation(PartitionSpec spec, StructLike partitionData, String filename) {
+      throw new IllegalStateException("Should have never been instantiated");
+    }
+  }
+
+  @Test
+  public void testDefaultLocationProvider() {
+    this.table.updateProperties()
+        .commit();
+
+    this.table.locationProvider().newDataLocation("my_file");
+    Assert.assertEquals(
+        "Default data path should have table location as root",
+        String.format("%s/data/%s", this.table.location(), "my_file"),
+        this.table.locationProvider().newDataLocation("my_file")
+    );
+  }
+
+  @Test
+  public void testDefaultLocationProviderWithCustomDataLocation() {
+    this.table.updateProperties()
+        .set(TableProperties.WRITE_NEW_DATA_LOCATION, "new_location")
+        .commit();
+
+    this.table.locationProvider().newDataLocation("my_file");
+    Assert.assertEquals(
+        "Default location provider should allow custom path location",
+        "new_location/my_file",
+        this.table.locationProvider().newDataLocation("my_file")
+    );
+  }
+
+  @Test
+  public void testDynamicallyLoadedLocationProvider() {
+    this.table.updateProperties()
+        .set(TableProperties.LOCATION_PROVIDER_IMPL,
+            String.format("%s$%s",
+                this.getClass().getCanonicalName(),
+                DynamicallyLoadedLocationProvider.class.getSimpleName()))
+        .commit();
+
+    Assert.assertTrue(String.format("Table should load impl defined in its properties"),
+        this.table.locationProvider() instanceof DynamicallyLoadedLocationProvider
+    );
+
+    Assert.assertEquals(
+        "Custom provider should take base table location",
+        String.format("%s/test_custom_provider/%s", this.table.location(), "my_file"),
+        this.table.locationProvider().newDataLocation("my_file")
+    );
+  }
+
+  @Test
+  public void testDynamicallyLoadedLocationProviderNotFound() {
+    String nonExistentImpl = String.format("%s$NonExistent%s",
+        this.getClass().getCanonicalName(),
+        DynamicallyLoadedLocationProvider.class.getSimpleName());
+    this.table.updateProperties()
+        .set(TableProperties.LOCATION_PROVIDER_IMPL, nonExistentImpl)
+        .commit();
+
+    exceptionRule.expect(IllegalArgumentException.class);
+    exceptionRule.expectMessage(
+        String.format("Unable to instantiate the provided implementation %s for %s.",
+            nonExistentImpl,
+            LocationProvider.class));
+    this.table.locationProvider();
+  }
+
+  @Test
+  public void testInvalidDynamicallyLoadedLocationProvider() {
```

Comment:
What happens if the class doesn't implement `LocationProvider`?

Could you add a test case for that?